### ISSUE TITLE
Segfault when calling numpy.linalg.svd on large enough arrays

### ISSUE THREAD

**Issue by ogrisel** [mingwpy/mingwpy#2]
Using the latest numpy package from pypi.anaconda.org/carlkl/simple it's possible to trigger a segfault with:

```
python -c "import numpy as np; X = np.random.randn(200, 200); print(np.linalg.svd(X))"
```

Note that with small dimensions and some other bigger values (e.g. 300x300) this can work without problem.

Note that `np.linalg.eig` and `eigh` can also trigger the segfault which makes me think that this is an issue with the OpenBLAS LAPACK bindings.

Note that the numpy tests do not segfault, which probably means that numpy tests only use small arrays. Here is the end of the test report on that host (VirtualBox VM with Windows 10 with Python 3.4.3 64 bit):

```
Ran 18291 tests in 313.516s

FAILED (KNOWNFAIL=98, SKIP=1193, errors=2, failures=4)
```

There are some failures but I think @carlkl already know about them.

**Comment by ogrisel** [comment]
Note that this happens with the `Sandybridge` core as detected by:

https://gist.github.com/ogrisel/ad4e547a32d0eb18b4ff

If I force the `Nehalem` platform by setting the following environment variable (powershell):

```
$env:OPENBLAS_CORETYPE="Nehalem"
```

then the crash disappears. So this is probably a problem in the dynamic architecture detection by openblas. It might be a consequence of the fact that I use VirtualBox to run windows.

**Comment by ogrisel** [comment]
Here is the output of https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/cpuid_features.exe on that host:

```
Vendor    = GenuineIntel
Processor =        Intel(R) Core(TM) i7-3667U CPU @ 2.00GHz

YES - FPU           (Floating-point Unit on-chip)
YES - VME           (Virtual Mode Extension)
YES - DE            (Debugging Extension)
YES - PSE           (Page Size Extension)
YES - TSC           (Time Stamp Counter)
YES - MSR           (Model Specific Registers)
YES - PAE           (Physical Address Extension)
YES - MCE           (Machine Check Exception)
YES - CX8           (CMPXCHG8 Instructions)
YES - APIC          (On-chip APIC hardware)
YES - SEP           (Fast System Call)
YES - MTRR          (Memory type Range Registers)
YES - PGE           (Page Global Enable)
YES - MCA           (Machine Check Architecture)
YES - CMOV          (Conditional Move Instruction)
YES - PAT           (Page Attribute Table)
YES - PSE36         (36bit Page Size Extension
NO  - PSN           (Processor Serial Number)
YES - CLFSH         (CFLUSH Instruction)
NO  - DS            (Debug Store)
NO  - ACPI          (Thermal Monitor & Software Controlled Clock)
YES - MMX           (Multi-Media Extension)
YES - FXSR          (Fast Floating Point Save & Restore)
YES - SSE           (Streaming SIMD Extension 1)
YES - SSE2          (Streaming SIMD Extension 2)
NO  - SS            (Self Snoop)
YES - HTT           (Hyper Threading Technology)
NO  - TM            (Thermal Monitor)
NO  - PBE           (Pend Break Enabled)
YES - SSE3          (Streaming SMD Extension 3)
NO  - MW            (Monitor Wait Instruction
NO  - CPL           (CPL-qualified Debug Store)
NO  - VMX           (Virtual Machine Extensions)
NO  - EST           (Enchanced Speed Test)
NO  - TM2           (Thermal Monitor 2)
YES - SSSE3         (Supplemental Streaming SIMD Extensions 3)
NO  - L1            (L1 Context ID)
NO  - FMA3          (Fused Multiply-Add 3-operand Form)
YES - CAE           (Compare And Exchange 16B)
YES - SSE41         (Streaming SIMD Extensions 4.1)
YES - SSE42         (Streaming SIMD Extensions 4.2)
YES - POPCNT        (Advanced Bit Manipulation - Bit Population Count Instruction)
YES - AES           (Advanced Encryption Standard)
YES - AVX           (Advanced Vector Extensions)
YES - RDRAND        (Random Number Generator)
NO  - AVX2          (Advanced Vector Extensions 2)
NO  - BMI1          (Bit Manipulations Instruction Set 1)
NO  - BMI2          (Bit Manipulations Instruction Set 2)
NO  - ADX           (Multi-Precision Add-Carry Instruction Extensions)
NO  - AVX512F       (512-bit extensions to Advanced Vector Extensions Foundation)
NO  - AVX512PFI     (512-bit extensions to Advanced Vector Extensions Prefetch Instructions)
NO  - AVX512ERI     (512-bit extensions to Advanced Vector Extensions Exponential and Reciprocal Instructions)
NO  - AVX512CDI     (512-bit extensions to Advanced Vector Extensions Conflict Detection Instructions)
NO  - SHA           (Secure Hash Algorithm)
YES - X64           (64-bit Extensions/Long mode)
NO  - LZCNT         (Advanced Bit Manipulation - Leading Zero Bit Count Instruction)
NO  - SSE4A         (Streaming SIMD Extensions 4a)
NO  - FMA4          (Fused Multiply-Add 4-operand Form)
NO  - XOP           (Extended Operations)
NO  - TBM           (Trailing Bit Manipulation Instruction)
NO  - LWP           (Light Weight Profiling Support)
NO  - WDT           (Watchdog Timer Support)
NO  - IBS           (Instruction Based Sampling)
NO  - 3DNOWPREFETCH (PREFETCH and PREFETCHW instruction support)
NO  - MISALIGNSSE   (Misaligned SSE mode)
NO  - SVM           (Secure Virtual Machine)
YES - LAHFSAHF      (LAHF and SAHF instruction support in 64-bit mode)
```

**Comment by carlkl** [comment]
is it the same or similar to as https://github.com/xianyi/OpenBLAS/issues/603?
@ogrisel, can you try the libopenblaspy.dll as drop in replacement taken from
https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/openblas_mingwpy-dev_8d6c8dc_amd64.7z
You can also test the wheels from https://anaconda.org/mingwpy/numpy instead of https://anaconda.org/carlkl/numpy

**Comment by ogrisel** [comment]
I don't think this is exactly the same issue as this not running on a Barcelona host (under VM) but on a Sandybridge host (under VM as well).

**Comment by carlkl** [comment]
Maybe it is due to the fact, that avx and avx2 is disabled in the virtualbox: https://www.virtualbox.org/wiki/Changelog? If that is true, explicit exporting  a non-avx target - NEHALEM - is mandantory. An openblas runtime check for this behaviour would we fine in this case.
**EDIT** maybe this one: https://github.com/xianyi/OpenBLAS/issues/124#issuecomment-6860090

**Comment by ogrisel** [comment]
Indeed but this has been addressed in 2013. There might be a related bug though. I think we should try to write a minimalistic C program to reproduce the issue and open a new issue upstream.

Note that I observed a similar issue under Linux (also under virtualbox via docker for OSX) with the manylinux1 wheel package for numpy and openblas (in [this thread on numpy-discussion](http://numpy-discussion.10968.n7.nabble.com/Multi-distribution-Linux-wheels-please-test-td42309.html)).

**Comment by ogrisel** [comment]
I forgot to report, I can reproduce the issue with the numpy / openblas from https://pypi.anaconda.org/mingwpy/simple as well.

**Comment by ogrisel** [comment]
I am running VirtualBox 5.0.14. AVX and AVX2 and XSAVE are supposedly supported since 5.0 beta3:

http://www.phoronix.com/scan.php?page=news_item&px=VirtualBox-5.0-Beta-3

but apparently AVX2 was disabled afterwards (in  5.0.2):

https://www.virtualbox.org/wiki/Changelog

Note that AVX instructions are reported by cpuid_features.exe but not AVX2 which matches the description. Sandybridge should only support AVX and not AVX2 so this is all fine apparently (unless AVX support is buggy in virtualbox).

**Comment by carlkl** [comment]
I now made some tests too (on a Haswell CPU)  and got segfaults with PRESCOTT, NORTHWOOD, SANDYBRIDGE targets but succeed with NEHALEM and HASWELL.

But it works for me on all targets with the following approach: cherrypicking libopenblaspy.dll from https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/openblas_mingwpy-dev_8d6c8dc_amd64.7z (just exchange libopenblaspy.dll in the numpy core folder)

**Comment by ogrisel** [comment]
Indeed using that version of the openblas dll fixed the issue on my VM as well.

However I now get the following error trying to do the same from scipy:

```
PS C:\Users\ogrisel> C:\Python34_orig\python -c "import numpy as np; from scipy import linalg; X = np.random.randn(200, 200); print(linalg.svd(X))"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Python34_orig\lib\site-packages\scipy\linalg\__init__.py", line 172, in <module>
    from .misc import *
  File "C:\Python34_orig\lib\site-packages\scipy\linalg\misc.py", line 6, in <module>
    from .lapack import get_lapack_funcs
  File "C:\Python34_orig\lib\site-packages\scipy\linalg\lapack.py", line 356, in <module>
    from scipy.linalg import _flapack
ImportError: DLL load failed: The specified procedure could not be found.
```

**Comment by ogrisel** [comment]
BTW I still think we should produce a minimalist C program to reproduce the original problem and report it upstream.

**Comment by carlkl** [comment]
Interesting scipy failure: `ImportError: DLL load failed: The specified procedure could not be found.`

This is due to the fact, that OpenBLAS now uses LAPACK 3.6.0 and symbols like `cgegv` are now marked deprecated in Lapack 

**quote** from http://www.netlib.org/lapack/lapack-3.6.0.html

_Improvements to nonsymmetric generalized eigenvalue problems xGGHD3: blocked Hessenberg-triangular reduction routines, xGGHD3, for nonsymmetric generalized eigenvalue problems. This replaces the current unblocked routines xGGHRD. CONTRIBUTION BY Daniel Kressner and Meiyue Shao_

```
    [Philippe Theveny, UC Denver] Deprecated routines

    CGEGS   CGEGV   CGELSX  CGEQPF  CGGSVD
    CGGSVP  CLAHRD  CLATZM  CTZRQF
    DGEGS   DGEGV   DGELSX  DGEQPF  DGGSVD
    DGGSVP  DLAHRD  DLATZM  DTZRQF
    SGEGS   SGEGV   SGELSX  SGEQPF  SGGSVD
    SGGSVP  SLAHRD  SLATZM  STZRQF
    ZGEGS   ZGEGV   ZGELSX  ZGEQPF  ZGGSVD
    ZGGSVP  ZLAHRD  ZLATZM  ZTZRQF
```

As some of this routines are still needed for scipy I will file a bug in OpenBLAS.

**EDIT**

https://bugzilla.redhat.com/show_bug.cgi?id=1287405
https://github.com/xianyi/OpenBLAS/issues/707

**Comment by rgommers** [comment]
All functions deprecated in LAPACK 3.6.0 were removed in scipy 0.17.0: https://github.com/scipy/scipy/blob/master/doc/release/0.17.0-notes.rst#backwards-incompatible-changes

So no bug filing needed I think, just use `>= 0.17.0`

**Comment by carlkl** [comment]
There is nothing that speaks against an openblas build that includes the deprecated symbols?

On the other hand it seems ok if scipy < 0.17.0 isn't suported.

I don't have a final position on that.

**Comment by rgommers** [comment]
> There is nothing that speaks against an openblas build that includes the deprecated symbols?

Indeed. If OpenBLAS wants to support those for longer, great. I think that that removal in LAPACK 3.6.0 was way too abrupt. I was just trying to say that it's not actually an OpenBLAS bug to follow the reference LAPACK.

**Comment by carlkl** [comment]
Lapack supports build of these symbols with an undocumented directive, that has to be included into a perl script used by the OpenBLAS build system as well. That's all.

**Comment by carlkl** [comment]
@rgommers, shouldn't https://github.com/scipy/scipy/issues/5266 be closed?

**Comment by rgommers** [comment]
Yes, it should. But the discussion at the end about `ggsvd3` needs moving to another issue and moving forward before doing so.

**Comment by carlkl** [comment]
@ogrisel, can you test the svd issue again with https://bitbucket.org/carlkl/mingw-w64-for-python/downloads/mingwpy-OpenBLAS-d643d48_amd64.7z.  It works for me regardless of the settings of `OPENBLAS_CORETYPE`

**EDIT**  build was done with:

```
git clone -b mingwpy-dev https://github.com/mingwpy/OpenBLAS.git
make BINARY=64 CC=gcc FC=gfortran VERSION=0.2.16.dev TARGET=NEHALEM DYNAMIC_ARCH=1 USE_THREAD=1 USE_OPENMP=0 NUM_THREADS=32 NO_WARMUP=1 NO_AFFINITY=1 CONSISTENT_FPCSR=1 BUILD_LAPACK_DEPRECATED=1 COMMON_OPT="-O2 -march=x86-64 -mtune=generic" FCOMMON_OPT="-O2 -march=x86-64 -mtune=generic -frecursive -ffpe-summary=invalid,zero"  MAX_STACK_ALLOC=2048
```

**Comment by ogrisel** [comment]
Sorry @carlkl for the late feedback. I tested with the new version of libopenblaspy.dll you provided in the previous comment and I can no longer reproduce this crash.

Now that openblas 0.2.17 is out, it might be interesting to make a new release of the experimental numpy  and scipy wheel with that version.

**Comment by carlkl** [comment]
thanks for doing this. I have a more recent OpenBLAS build, that is equivalent to 0.2.17 (only non-relevant commits are missing)
**EDIT**: OpenBLAS-0.2.17 binaries now available at https://bitbucket.org/carlkl/mingw-w64-for-python/downloads
